### ISSUE TITLE
CRF formula wrong?

### ISSUE THREAD

**Issue by helpmefindaname** [keras-team/keras-contrib#460]
Looking at https://github.com/keras-team/keras-contrib/blob/fff264273d5347613574ff533c598f55f15d4763/keras_contrib/layers/crf.py#L470 and https://github.com/keras-team/keras-contrib/blob/fff264273d5347613574ff533c598f55f15d4763/keras_contrib/layers/crf.py#L474
we see two almost identical lines, calculating the energy. However one add's `prev_target_val` while the other substracts `prev_target_val`.
I don't understand the formula completely, but I assume those two lines should be the same. Can someone please verify how the formula is done right or explain how the different sign suits the formula?

**Comment by gabrieldemarmiesse** [comment]
@lzfelix would you happen to know something about that?

**Comment by lzfelix** [comment]
I am a bit sick these days, but I'll take a look as soon as I can.

**Comment by gabrieldemarmiesse** [comment]
@lzfelix no worries, I wish you a quick recorevery. Thanks for all the help.

**Comment by LCorleone** [comment]
@lzfelix @gabrieldemarmiesse 
I have the same question, is there any ideas? thanks!

**Comment by helpmefindaname** [comment]
From digging more into it, I noticed that I didn't take a proper enough look at it.
In https://github.com/keras-team/keras-contrib/blob/fff264273d5347613574ff533c598f55f15d4763/keras_contrib/layers/crf.py#L471 we see a second minus.
So the first calculation uses two negatives, while the second has none. 
In the first calculation the softmax of the negative input energy+chain energy is taken so small numbers -> higher probability
In the second, the smallest positive input energy + chain energy is taken -> element with highest probability
So all in all it is consistent.

The only thing that I find strange is that the chain energy is the same for the forward and backward pass.  chain energy ~ log(P(y_i, y_{i+1}). In my opinion it should be better to transpose it on the backward pass because  log(P(y_i, y_{i+1}) !=  log(P(y_{i+1}, y_i)
